benchmark/src/main/java/com/datatorrent/benchmark/window/KeyedWindowedOperatorBenchmarkApp.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.apex.malhar.lib.window.WindowedStorage;
+import org.apache.apex.malhar.lib.window.accumulation.Count;
+import org.apache.apex.malhar.lib.window.impl.InMemoryWindowedKeyedStorage;
+import org.apache.apex.malhar.lib.window.impl.KeyedWindowedOperatorImpl;
+import org.apache.apex.malhar.lib.window.impl.SpillableWindowedKeyedStorage;
+
+import com.datatorrent.api.DAG;
+import com.datatorrent.api.DAG.Locality;
+import com.datatorrent.api.DefaultOutputPort;
+import com.datatorrent.lib.util.KeyValPair;
+
+public class KeyedWindowedOperatorBenchmarkApp extends AbstractWindowedOperatorBenchmarkApp<KeyedWindowedOperatorBenchmarkApp.KeyedWindowedGenerator, KeyedWindowedOperatorBenchmarkApp.MyKeyedWindowedOperator>
+{
+  public KeyedWindowedOperatorBenchmarkApp()
+  {
+    generatorClass = KeyedWindowedGenerator.class;
+    windowedOperatorClass = KeyedWindowedOperatorBenchmarkApp.MyKeyedWindowedOperator.class;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  protected void connectGeneratorToWindowedOperator(DAG dag, KeyedWindowedGenerator generator,
+      KeyedWindowedOperatorBenchmarkApp.MyKeyedWindowedOperator windowedOperator)
+  {
+    dag.addStream("Data", generator.data, windowedOperator.input).setLocality(Locality.CONTAINER_LOCAL);
+  }
+
+  protected static class MyKeyedWindowedOperator extends KeyedWindowedOperatorImpl
+  {
+    private static final Logger logger = LoggerFactory.getLogger(MyKeyedWindowedOperator.class);
+
+    private long logWindows = 20;
+    private long windowCount = 0;
+    private long beginTime = System.currentTimeMillis();
+    private long tupleCount = 0;
+    private long totalBeginTime = System.currentTimeMillis();
+    private long totalCount = 0;
+
+    private long droppedCount = 0;
+    @Override
+    public void dropTuple(Tuple input)
+    {
+      droppedCount++;
+    }
+
+    @Override
+    public void endWindow()
+    {
+      super.endWindow();
+      if (++windowCount == logWindows) {
+        long endTime = System.currentTimeMillis();
+        tupleCount -= droppedCount;
+        totalCount += tupleCount;
+        logger.info("total: count: {}; time: {}; average: {}; period: count: {}; dropped: {}; time: {}; average: {}",
+            totalCount, endTime - totalBeginTime, totalCount * 1000 / (endTime - totalBeginTime),
+            tupleCount, droppedCount, endTime - beginTime, tupleCount * 1000 / (endTime - beginTime));
+        windowCount = 0;
+        beginTime = System.currentTimeMillis();
+        tupleCount = 0;
+        droppedCount = 0;
+      }
+    }
+
+    @Override
+    public void processTuple(Tuple tuple)
+    {
+      super.processTuple(tuple);
+      ++tupleCount;
+    }
+  }
+
+  protected static class KeyedWindowedGenerator extends AbstractGenerator
+  {
+    public final transient DefaultOutputPort<Tuple.TimestampedTuple<KeyValPair<String, Long>>> data = new DefaultOutputPort<Tuple.TimestampedTuple<KeyValPair<String, Long>>>();
```

Comment:
Currently, the AbstractGenerator actually does not have anything that has to do with "generating". I think we should have the AbstractGenerator class take the output type parameter and declare the data port there, and have the emitTuples be inside the abstract class. The concrete class only provide the tuple. Like this:
```java
public abstract class AbstractGenerator<T> extends BaseOperator implements InputOperator
{
  ...
  public final transient DefaultOutputPort<T> data = new DefaultOutputPort<>(); 
  ...
  @Override
  public void emitTuples()
  {
    for (int i = 0; i < emitBatchSize && emitCount < rate; i++) {
      data.emit(generateTuple());
      emitCount++;
    }
  }
  public abstract T generateTuple();
}
```
The rest should be obvious.